File: javascript/snippets/rover.js
var DIRECTIONS = {
    NORTH: 0,
    EAST: 1,
    SOUTH: 2,
    WEST: 3
};

var MOVEMENT_SPEED = 1;
var GRID_WIDTH = 5;
var GRID_HEIGHT = 5;

var rover = {};
var directionDisplayMapping = ['N', 'E', 'S', 'W'];
var movements = [[0, MOVEMENT_SPEED], [MOVEMENT_SPEED, 0], [0, -1 * MOVEMENT_SPEED], [-1 * MOVEMENT_SPEED, 0]];

function setGrid(config) {
    if (!config || typeof config !== 'string') return;
    const splitConfig = config.split(' ');
    if (splitConfig.length < 2) return;
    const [gridWidth, gridHeight] = splitConfig;
    GRID_WIDTH = gridWidth;
    GRID_HEIGHT = gridHeight;
}

function setRover(config) {
    if (!config || typeof config !== 'string') return;
    const splitConfig = config.split(' ');
    if (splitConfig.length < 3) return;
    const [startingX, startingY, startingDirection] = splitConfig;
    return { 
        x: Number(startingX) || 0, 
        y: Number(startingY) || 0, 
        direction: directionDisplayMapping.indexOf(startingDirection) || 0
    };
}

function parseInstructions(instructions) {
    if (!instructions || typeof instructions !== 'string') return;
    instructions.split('').forEach(instruction => {
        switch (instruction) {
            case 'M':
                const [xDelta,yDelta] = movements[rover.direction];
                if (rover.x + xDelta <= GRID_WIDTH && rover.x + xDelta >= 0) rover.x += xDelta;
                if (rover.y + yDelta <= GRID_HEIGHT && rover.y + yDelta >= 0) rover.y += yDelta;
                break;
            case 'L':
            case 'R':
                const rotation = instruction === 'R' ? 1 : -1;
                const newDirection = rover.direction + rotation;
                rover.direction = newDirection < 0 
                    ? movements.length - 1 
                    : newDirection % movements.length;           
                break;
        }
    });

    console.log(rover.x, rover.y, directionDisplayMapping[rover.direction]);
}

setGrid('5 5');
rover = setRover('1 2 N');
parseInstructions('LMLMLMLMM');

rover = setRover('3 3 E');
parseInstructions('MMRMMRMRRM');

File: ruby/rover/lib/rover.rb
DIRECTIONS = %i[N E S W]
DELTAS = {
  :N => [0, -1],
  :E => [1, 0],
  :S => [0, 1],
  :W => [-1, 0],
}

class Rover
  attr_accessor :grid_height, :grid_width, :x, :y, :direction

  def initialize
    @grid_height = 0
    @grid_width = 0
    @x = 0
    @y = 0
    self
  end

  def get_current_position
    [@x, @y, @direction.to_s]
  end

  def set_grid(grid_config)
    width, height = grid_config.split(" ")
    @grid_height = height.to_i
    @grid_width = width.to_i
    self
  end

  def set_starting_position(starting_position)
    starting_x, starting_y, starting_direction = starting_position.split(" ")
    puts "Setting starting position to #{starting_x} #{starting_y} facing #{starting_direction}..."
    @x = starting_x.to_i
    @y = starting_y.to_i
    @direction = starting_direction.intern
    self
  end

  def parse_instructions(instructions)
    instructions.chars.each do |instruction|
      case instruction
      when "M"
        puts "Moving forward..."
        x_delta, y_delta = DELTAS[@direction]
        @x += x_delta if 0 <= @x + x_delta and @x + x_delta <= @grid_width
        @y += y_delta if 0 <= @y + y_delta and @y + y_delta <= @grid_height
      when "L"
        puts "Turning left..."
        @direction = DIRECTIONS.rotate!(-1).first
      when "R"
        puts "Turning right..."
        @direction = DIRECTIONS.rotate!(1).first
      end
    end
    self
  end
end
